src/builtins/set.rs

```diff
@@ -1012,7 +1012,7 @@ pub fn set(parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> Opti
     };
 
     if retval == STATUS_CMD_OK && opts.preserve_failure_exit_status {
-        return None;
+        return STATUS_PRESERVE_FAILURE;
```

Comment:
This is the crux - the reason that there's so many Options everywhere is because `builtin set` should not modify the exit status if it succeeds. We should not reserve a value like `STATUS_PRESERVE_FAILURE` because then that becomes unavailable as the exit status of other commands.